src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/RazorDynamicDocumentInfoProviderBase.cs

```diff
@@ -0,0 +1,216 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Composition;
+using System.IO;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+
+namespace Microsoft.CodeAnalysis.Razor
+{
+    /// <summary>
+    /// Base MEF component for implementing a dynamic document info provider
+    /// Currently only used by VS Mac
+    /// </summary>
+    internal abstract class RazorDynamicDocumentInfoProviderBase
+    {
+        private readonly ConcurrentDictionary<Key, Entry> _entries;
+
+        [Import]
+        private DocumentServiceProviderFactory Factory { get; set; }
+
+        public RazorDynamicDocumentInfoProviderBase()
+        {
+            _entries = new ConcurrentDictionary<Key, Entry>();
+        }
+
+        public event Action<DocumentInfo> Updated;
```

Comment:
@KirillOsenkov can this return `string`? That's what the VS IDynamicFileInfo type expects

Reply:
Unfortunately we need to pass the ProjectId, the DocumentId and the actual DocumentInfo to update the workspace: https://github.com/mono/monodevelop/blob/c740bfab1077385701ebf3486970fef4d9e44a86/main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/DynamicFileManager.cs#L88-L97

Unlike the VSWin implementation we do not callback into RazorDynamicFileInfoProvider immediately because a DocumentInfo is all we really need.

If you're thinking about unifying the implementations again, at least partially, we'd be open to changes on the MonoDevelop side of things I guess...

Reply:
BTW here's the commit that adds support on the MonoDevelop side: https://github.com/mono/monodevelop/commit/b2c458d77565b79eb9c95110af69eec1495882eb

Reply:
Hmm, ya do you think we could change MonoDevelop's side to understand a string and then lookup the document/project info based on the string? I'd love to have these two pieces of code unified

Reply:
I think it sounds reasonable... wanna send a PR? You'll probably have to do more state tracking on the IDE side to be able to find the ProjectIds and DocumentIds for a file path, and then get the new DocumentInfo. The advantage of the current solution is that DocumentInfo neatly passes everything in a single package, including the loader, service provider, etc.

Reply:
If only I had more time 😢. Alright sounds like i'll need to scrap the unification idea.